cql/src/test/java/org/apache/cassandra/stargate/metrics/ClientMetricsTest.java

```diff
@@ -0,0 +1,486 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.cassandra.stargate.metrics;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.search.MeterNotFoundException;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import io.stargate.db.ClientInfo;
+import io.stargate.db.metrics.api.ClientInfoMetricsTagProvider;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.cassandra.stargate.transport.internal.Server;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class ClientMetricsTest {
+
+  ClientMetrics clientMetrics = ClientMetrics.instance;
+
+  MeterRegistry meterRegistry = new SimpleMeterRegistry();
+
+  // manual mocks, init before all
+  ClientInfoMetricsTagProvider clientTagProvider;
+  Server server1;
+  Server server2;
+  ClientInfo clientInfo1;
+  ClientInfo clientInfo2;
+
+  @BeforeAll
+  public void initMocks() {
+    clientTagProvider = mock(ClientInfoMetricsTagProvider.class);
+    server1 = mock(Server.class);
+    server2 = mock(Server.class);
+    clientInfo1 = mock(ClientInfo.class);
+    clientInfo2 = mock(ClientInfo.class);
+
+    when(clientTagProvider.getClientInfoTags(clientInfo1)).thenReturn(Tags.of("client", "one"));
+    when(clientTagProvider.getClientInfoTags(clientInfo2)).thenReturn(Tags.of("client", "two"));
+
+    List<Server> servers = Arrays.asList(server1, server2);
+    clientMetrics.init(servers, meterRegistry, clientTagProvider, 0);
+  }
+
+  @Nested
```

Comment:
why do we need those nested classes with test cases `happyPath()`? I think that the existing pattern for tests is to have it in a more flat structure, without nesting, AND having more descriptive names.

Reply:
I am a huge fan of Nested tests as there are so many advantages over the flat structure you are proposing to follow here. Let me explain:

First, if you open my test you can clearly see from the nested class name what (public) method is being tested in the cut. Meaning you have clear separation for target tested code/path. Open any flat structure test and you'll need to dig deep to actually figure out what's being tested in each test. And descriptive names don't help here imo, this is unit test, you are testing a peace of code no behavior.

Second, since you have that nested structure per the method, you can then actually have test methods in the nested class to handle different paths / situations. This gives you a clear view on what's covered for each method.

Third, once you have such a clear structure, adding new tests is easy. If you extended a existing method, add a new test method in the target nested class. If you added new method, add new nested class.

And, the last results outputs in the IDE and reports are we structured as well. Clear and readable, you can even directly associate the possible failing test with the peace of code that might influence it:

![image](https://user-images.githubusercontent.com/10600041/118104153-0cddbc00-b3db-11eb-9f10-cfb6a3caa948.png)

Reply:
Yeah, I agree that this IDE structure looks very clear!
However, IDE is one side of the coin. Is it also so user-friendly when running the `mvn` test runner?

Reply:
Neat, that structuring does look clean.

Regarding `happyPath` I'm personally cool with it and use it as well. It's a pretty common naming convention for testing in general

https://en.wikipedia.org/wiki/Happy_path